flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTaskNetworkInput.java

```diff
@@ -159,10 +158,12 @@ public InputStatus emitNext(DataOutput<T> output) throws Exception {
 			if (bufferOrEvent.isPresent()) {
 				// return to the mailbox after receiving a checkpoint barrier to avoid processing of
 				// data after the barrier before checkpoint is performed for unaligned checkpoint mode
-				if (bufferOrEvent.get().isEvent() && bufferOrEvent.get().getEvent() instanceof CheckpointBarrier) {
+				if (bufferOrEvent.get().isBuffer()) {
+					processBuffer(bufferOrEvent.get());
+				} else {
+					processEvent(bufferOrEvent.get());
```

Comment:
Nope, it's a refactoring that I should move out. We could also revert back. I had an intermediate version where `processEvent` signaled if the loop should be broken or not and there the split made more sense.

Reply:
Sorry I haven't pulled that change out, since it's more than a refactoring; it's the core change that modifies the semantics to always return on events. 
Except for that change, everything else is more or less directly related to passing the priority future, so I'm not sure what to pull out.